File: apps/worker/fetchers/ecb.py
"""Fetch daily exchange rates from ECB and insert into Postgres.

Fetches EUR-based rates from the ECB SDMX REST API, converts to
base_currency/quote_currency/rate pairs (quote_currency=USD),
and inserts missing dates into exchange_rates.
"""

import csv
import io
import logging
from datetime import date, timedelta
from decimal import Decimal, InvalidOperation
from typing import TypedDict

import psycopg2.extensions
import requests

from config import ECB_BASE_URL, ECB_CURRENCIES

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Type definitions
# ---------------------------------------------------------------------------

class ExchangeRateRow(TypedDict):
    base_currency: str
    quote_currency: str
    rate_date: str
    rate: str


class ECBRate(TypedDict):
    currency: str
    rate_date: str
    rate_eur: Decimal


class DateRange(TypedDict):
    min_date: date
    max_date: date


# ---------------------------------------------------------------------------
# Pure functions — ECB data parsing
# ---------------------------------------------------------------------------

def parse_ecb_csv(csv_text: str) -> list[ECBRate]:
    """Parse ECB CSV response into a list of rates.

    ECB CSV columns include CURRENCY, TIME_PERIOD, OBS_VALUE among others.
    Each row is one daily rate: how many units of CURRENCY per 1 EUR.
    """
    reader = csv.DictReader(io.StringIO(csv_text))
    rates: list[ECBRate] = []
    for row in reader:
        currency = row["CURRENCY"]
        time_period = row["TIME_PERIOD"]
        obs_value = row["OBS_VALUE"]
        if not obs_value:
            continue
        try:
            rate_eur = Decimal(obs_value)
        except InvalidOperation as exc:
            raise ValueError(
                f"Invalid rate value from ECB: currency={currency} "
                f"date={time_period} value={obs_value!r}"
            ) from exc
        rates.append(ECBRate(currency=currency, rate_date=time_period, rate_eur=rate_eur))
    return rates


def convert_eur_rates_to_usd(ecb_rates: list[ECBRate]) -> list[ExchangeRateRow]:
    """Convert EUR-based ECB rates to base_currency/USD pairs.

    ECB gives: 1 EUR = X units of CCY (rate_eur_ccy)
    ECB gives: 1 EUR = Y USD (rate_eur_usd)

    We produce rows with quote_currency=USD:
    - EUR→USD: rate = rate_eur_usd
    - CCY→USD: rate = rate_eur_usd / rate_eur_ccy
    """
    rates_by_date: dict[str, dict[str, Decimal]] = {}
    for rate in ecb_rates:
        rates_by_date.setdefault(rate["rate_date"], {})[rate["currency"]] = rate["rate_eur"]

    rows: list[ExchangeRateRow] = []
    for rate_date, day_rates in sorted(rates_by_date.items()):
        if "USD" not in day_rates:
            raise ValueError(
                f"No EUR/USD rate from ECB for {rate_date}. "
                f"Available currencies: {sorted(day_rates.keys())}"
            )
        eur_usd = day_rates["USD"]

        for currency, rate_eur in sorted(day_rates.items()):
            if currency == "USD":
                rows.append(ExchangeRateRow(
                    base_currency="EUR",
                    quote_currency="USD",
                    rate_date=rate_date,
                    rate=str(eur_usd),
                ))
            else:
                rate_to_usd = (eur_usd / rate_eur).quantize(Decimal("0.000000001"))
                rows.append(ExchangeRateRow(
                    base_currency=currency,
                    quote_currency="USD",
                    rate_date=rate_date,
                    rate=str(rate_to_usd),
                ))

    return rows


# ---------------------------------------------------------------------------
# Postgres operations
# ---------------------------------------------------------------------------

def get_rate_date_ranges(
    conn: psycopg2.extensions.connection,
    currencies: list[str],
) -> dict[str, DateRange]:
    """Query min and max rate_date per base_currency already in Postgres."""
    with conn.cursor() as cur:
        cur.execute(
            "SELECT base_currency, MIN(rate_date) AS min_date, MAX(rate_date) AS max_date "
            "FROM exchange_rates "
            "WHERE base_currency = ANY(%s) AND quote_currency = 'USD' "
            "GROUP BY base_currency",
            (currencies,),
        )
        return {
            row[0]: DateRange(min_date=row[1], max_date=row[2])
            for row in cur.fetchall()
        }


def get_earliest_transaction_date(conn: psycopg2.extensions.connection) -> date:
    """Query the earliest transaction date from ledger_entries."""
    with conn.cursor() as cur:
        cur.execute("SELECT MIN(ts::date) AS min_date FROM ledger_entries")
        row = cur.fetchone()
        if row and row[0] is not None:
            return row[0]
    return date.today() - timedelta(days=30)


def insert_rows(
    conn: psycopg2.extensions.connection,
    rows: list[ExchangeRateRow],
) -> int:
    """Insert rows into Postgres. Returns count of inserted rows."""
    if not rows:
        return 0
    with conn.cursor() as cur:
        values = [
            (r["base_currency"], r["quote_currency"], r["rate_date"], r["rate"])
            for r in rows
        ]
        cur.executemany(
            "INSERT INTO exchange_rates (base_currency, quote_currency, rate_date, rate) "
            "VALUES (%s, %s, %s, %s) "
            "ON CONFLICT (base_currency, quote_currency, rate_date) DO NOTHING",
            values,
        )
        conn.commit()
    return len(rows)


# ---------------------------------------------------------------------------
# ECB API
# ---------------------------------------------------------------------------

def fetch_ecb_rates(currencies_with_usd: list[str], start_period: str, end_period: str) -> str:
    """Fetch daily rates from ECB SDMX REST API as CSV."""
    currency_key = "+".join(currencies_with_usd)
    url = f"{ECB_BASE_URL}/D.{currency_key}.EUR.SP00.A"
    params = {"format": "csvdata", "startPeriod": start_period, "endPeriod": end_period}

    logger.info(
        "Fetching ECB rates",
        extra={"url": url, "params": params},
    )

    response = requests.get(url, params=params, timeout=30)
    response.raise_for_status()
    return response.text


# ---------------------------------------------------------------------------
# Orchestration
# ---------------------------------------------------------------------------

def determine_ecb_currencies(requested: list[str]) -> list[str]:
    """Build the currency list for the ECB API call.

    ECB rates are EUR-based. We always need USD in the response to convert
    other currencies to USD. For EUR itself, we only need the USD rate.
    Non-EUR/USD currencies are fetched directly.
    """
    ecb_currencies: set[str] = {"USD"}
    for ccy in requested:
        if ccy == "USD":
            raise ValueError("USD should not be in ECB_CURRENCIES config (it's the target)")
        if ccy != "EUR":
            ecb_currencies.add(ccy)
    return sorted(ecb_currencies)


def filter_new_rows(
    all_rows: list[ExchangeRateRow],
    date_ranges: dict[str, DateRange],
    requested_currencies: list[str],
) -> list[ExchangeRateRow]:
    """Keep only rows not already covered by existing data."""
    new_rows: list[ExchangeRateRow] = []
    for row in all_rows:
        if row["base_currency"] not in requested_currencies:
            continue
        row_date = date.fromisoformat(row["rate_date"])
        existing = date_ranges.get(row["base_currency"])
        if existing is None or row_date < existing["min_date"] or row_date > existing["max_date"]:
            new_rows.append(row)
    return new_rows


def run(conn: psycopg2.extensions.connection) -> dict[str, str | int | list[str]]:
    """Main logic: fetch missing rates from ECB and insert into Postgres."""
    ecb_currencies = determine_ecb_currencies(ECB_CURRENCIES)
    date_ranges = get_rate_date_ranges(conn, ECB_CURRENCIES)
    target_start = get_earliest_transaction_date(conn)

    all_currencies_present = all(c in date_ranges for c in ECB_CURRENCIES)
    needs_backfill = (
        not all_currencies_present
        or any(r["min_date"] > target_start for r in date_ranges.values())
    )

    if needs_backfill:
        start = target_start
    else:
        earliest_max = min(r["max_date"] for r in date_ranges.values())
        start = earliest_max + timedelta(days=1)

    end = date.today()

    if start > end:
        logger.info("All ECB rates are up to date")
        return {"inserted": 0, "latest_date": str(end), "missing_currencies": []}

    csv_text = fetch_ecb_rates(ecb_currencies, str(start), str(end))
    ecb_rates = parse_ecb_csv(csv_text)
    all_rows = convert_eur_rates_to_usd(ecb_rates)

    returned_currencies = {r["base_currency"] for r in all_rows}
    missing: list[str] = [c for c in ECB_CURRENCIES if c not in returned_currencies]
    if missing:
        logger.warning(
            "ECB did not return rates for currencies: %s (they may be suspended)",
            missing,
        )

    new_rows = filter_new_rows(all_rows, date_ranges, ECB_CURRENCIES)
    inserted = insert_rows(conn, new_rows)

    latest_inserted = max((r["rate_date"] for r in new_rows), default=str(end))

    logger.info("ECB: inserted %d rows, latest date: %s", inserted, latest_inserted)

    return {
        "inserted": inserted,
        "latest_date": latest_inserted,
        "missing_currencies": missing,
    }

File: apps/worker/src/fetchers/ecb.ts
/**
 * Fetch daily exchange rates from ECB and insert into Postgres.
 *
 * Fetches EUR-based rates from the ECB SDMX REST API, converts to
 * base_currency/quote_currency/rate pairs (quote_currency=USD),
 * and inserts missing dates into fx_rates_raw.
 */

import { ECB_BASE_URL, ECB_CURRENCIES, ECB_EARLIEST_DATE } from "../config";
import { addDays, daysBetween, todayIso } from "../dateUtils";
import { getRateDateRanges, insertRows } from "../dbQueries";
import type { FxRawRateRow, DateRange, FetcherResult } from "../types";

const SOURCE = "ecb";

// ---------------------------------------------------------------------------
// Type definitions
// ---------------------------------------------------------------------------

interface ECBRate {
  currency: string;
  rate_date: string;
  rate_eur: number;
}

// ---------------------------------------------------------------------------
// Pure functions — ECB data parsing
// ---------------------------------------------------------------------------

/**
 * Parse ECB CSV response into a list of rates.
 *
 * ECB CSV columns include CURRENCY, TIME_PERIOD, OBS_VALUE among others.
 * Each row is one daily rate: how many units of CURRENCY per 1 EUR.
 *
 * A period that contains no observations — a weekend, a TARGET holiday, or the
 * hours before the ~16:00 CET publication — is answered with HTTP 200 and a
 * completely empty body, not with a header-only CSV, so that is zero rates.
 */
function parseEcbCsv(csvText: string): ECBRate[] {
  if (csvText.trim() === "") {
    return [];
  }

  const lines = csvText.split("\n");
  const headers = lines[0].split(",");
  const currencyIdx = headers.indexOf("CURRENCY");
  const timePeriodIdx = headers.indexOf("TIME_PERIOD");
  const obsValueIdx = headers.indexOf("OBS_VALUE");

  if (currencyIdx === -1 || timePeriodIdx === -1 || obsValueIdx === -1) {
    throw new Error(
      `ECB CSV missing required columns. Headers: ${headers.join(", ")}`,
    );
  }

  const rates: ECBRate[] = [];
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;
    const cols = line.split(",");
    const currency = cols[currencyIdx];
    const timePeriod = cols[timePeriodIdx];
    const obsValue = cols[obsValueIdx];
    if (!obsValue) continue;
    const rateEur = Number(obsValue);
    if (Number.isNaN(rateEur)) {
      throw new Error(
        `Invalid rate value from ECB: currency=${currency} date=${timePeriod} value=${JSON.stringify(obsValue)}`,
      );
    }
    rates.push({ currency, rate_date: timePeriod, rate_eur: rateEur });
  }
  return rates;
}

/**
 * Convert EUR-based ECB rates to base_currency/USD pairs.
 *
 * ECB gives: 1 EUR = X units of CCY (rate_eur_ccy)
 * ECB gives: 1 EUR = Y USD (rate_eur_usd)
 *
 * We produce rows with quote_currency=USD:
 * - EUR->USD: rate = rate_eur_usd
 * - CCY->USD: rate = rate_eur_usd / rate_eur_ccy
 */
function convertEurRatesToUsd(ecbRates: ECBRate[]): FxRawRateRow[] {
  const ratesByDate: Record<string, Record<string, number>> = {};
  for (const rate of ecbRates) {
    if (!ratesByDate[rate.rate_date]) {
      ratesByDate[rate.rate_date] = {};
    }
    ratesByDate[rate.rate_date][rate.currency] = rate.rate_eur;
  }

  const rows: FxRawRateRow[] = [];
  for (const rateDate of Object.keys(ratesByDate).sort()) {
    const dayRates = ratesByDate[rateDate];
    if (dayRates["USD"] === undefined) {
      throw new Error(
        `No EUR/USD rate from ECB for ${rateDate}. Available currencies: ${Object.keys(dayRates).sort().join(", ")}`,
      );
    }
    const eurUsd = dayRates["USD"];

    for (const currency of Object.keys(dayRates).sort()) {
      if (currency === "USD") {
        rows.push({
          base_currency: "EUR",
          quote_currency: "USD",
          rate_date: rateDate,
          rate: eurUsd.toFixed(9),
          source: SOURCE,
        });
      } else {
        const rateToUsd = eurUsd / dayRates[currency];
        rows.push({
          base_currency: currency,
          quote_currency: "USD",
          rate_date: rateDate,
          rate: rateToUsd.toFixed(9),
          source: SOURCE,
        });
      }
    }
  }
  return rows;
}

// ---------------------------------------------------------------------------
// ECB API
// ---------------------------------------------------------------------------

// ECB answers one request for the whole requested range, and its cost is server-side
// query time rather than transfer: a one-year range is measured at 4-16 s and the full
// 1999-to-today history at ~122 s, almost all of it before the first byte. So the
// per-attempt timeout grows with the requested range instead of being a single fixed
// value, and the whole retry loop is bounded by ECB_TOTAL_BUDGET_MS, which leaves room
// inside the 300 s Lambda for the other five fetchers (~1 s) and the daily rebuild (~13 s).
const ECB_REQUEST_TIMEOUT_BASE_MS = 15_000;

const ECB_REQUEST_TIMEOUT_PER_DAY_MS = 20;

const ECB_REQUEST_TIMEOUT_MAX_MS = 240_000;

const ECB_TOTAL_BUDGET_MS = 240_000;

const ECB_REQUEST_ATTEMPTS = 3;

const ECB_RETRY_DELAY_MS = 1_000;

/** One HTTP attempt, classified into what the caller should do next. */
type EcbAttempt =
  | { kind: "ok"; body: string }
  | { kind: "transient"; error: Error }
  | { kind: "permanent"; error: Error };

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Per-attempt timeout for a request covering `days` days.
 *
 * The ceiling stays inside the Lambda budget, so a full-history range still gets one
 * complete attempt, while a short daily range keeps its full retry allowance.
 */
function requestTimeoutMs(days: number): number {
  return Math.min(
    ECB_REQUEST_TIMEOUT_BASE_MS + days * ECB_REQUEST_TIMEOUT_PER_DAY_MS,
    ECB_REQUEST_TIMEOUT_MAX_MS,
  );
}

async function attemptEcbRequest(url: string, timeoutMs: number): Promise<EcbAttempt> {
  try {
    const response = await fetch(url, { signal: AbortSignal.timeout(timeoutMs) });
    if (response.ok) {
      return { kind: "ok", body: await response.text() };
    }
    // 5xx and 429 can clear on their own; any other 4xx is a defect in the request
    // itself (bad currency key, malformed period) that an identical retry repeats.
    const error = new Error(`ECB API error: ${response.status} ${response.statusText}`);
    if (response.status >= 500 || response.status === 429) {
      return { kind: "transient", error };
    }
    return { kind: "permanent", error };
  } catch (error) {
    // Connection failure, or the per-attempt timeout aborting the request or body read.
    return {
      kind: "transient",
      error: error instanceof Error ? error : new Error(String(error)),
    };
  }
}

/**
 * Request the ECB CSV, retrying only transient failures.
 *
 * Attempt 1 always runs; a further one is started only while the elapsed time plus one
 * more timeout still fits ECB_TOTAL_BUDGET_MS, so the loop stays within that budget.
 */
async function fetchEcbCsv(url: string, timeoutMs: number): Promise<string> {
  const deadline = Date.now() + ECB_TOTAL_BUDGET_MS;
  let lastError: Error | undefined;

  for (let attempt = 1; attempt <= ECB_REQUEST_ATTEMPTS; attempt++) {
    if (attempt > 1) {
      await delay(ECB_RETRY_DELAY_MS * (attempt - 1));
      if (Date.now() + timeoutMs > deadline) {
        break;
      }
    }

    const result = await attemptEcbRequest(url, timeoutMs);
    if (result.kind === "ok") {
      return result.body;
    }
    if (result.kind === "permanent") {
      throw result.error;
    }

    lastError = result.error;
    console.warn("ECB request attempt failed", {
      url,
      attempt,
      attempts: ECB_REQUEST_ATTEMPTS,
      timeout_ms: timeoutMs,
      error: result.error.message,
    });
  }

  throw lastError ?? new Error(`ECB request failed after ${ECB_REQUEST_ATTEMPTS} attempts: ${url}`);
}

/** Fetch daily rates from ECB SDMX REST API as CSV. */
async function fetchEcbRates(
  currenciesWithUsd: string[],
  startPeriod: string,
  endPeriod: string,
): Promise<string> {
  const currencyKey = currenciesWithUsd.join("+");
  const url = `${ECB_BASE_URL}/D.${currencyKey}.EUR.SP00.A?format=csvdata&startPeriod=${startPeriod}&endPeriod=${endPeriod}`;
  const days = daysBetween(startPeriod, endPeriod) + 1;
  const timeoutMs = requestTimeoutMs(days);

  console.log("Fetching ECB rates", { url, days, timeout_ms: timeoutMs });

  return fetchEcbCsv(url, timeoutMs);
}

// ---------------------------------------------------------------------------
// Orchestration
// ---------------------------------------------------------------------------

/**
 * Build the currency list for the ECB API call.
 *
 * ECB rates are EUR-based. We always need USD in the response to convert
 * other currencies to USD. For EUR itself, we only need the USD rate.
 * Non-EUR/USD currencies are fetched directly.
 */
function determineEcbCurrencies(requested: string[]): string[] {
  const ecbCurrencies = new Set<string>(["USD"]);
  for (const ccy of requested) {
    if (ccy === "USD") {
      throw new Error("USD should not be in ECB_CURRENCIES config (it's the target)");
    }
    if (ccy !== "EUR") {
      ecbCurrencies.add(ccy);
    }
  }
  return Array.from(ecbCurrencies).sort();
}

/** Keep only rows not already covered by existing data. */
function filterNewRows(
  allRows: FxRawRateRow[],
  dateRanges: Record<string, DateRange>,
  requestedCurrencies: string[],
): FxRawRateRow[] {
  const newRows: FxRawRateRow[] = [];
  for (const row of allRows) {
    if (!requestedCurrencies.includes(row.base_currency)) {
      continue;
    }
    const existing = dateRanges[row.base_currency];
    if (!existing || row.rate_date < existing.min_date || row.rate_date > existing.max_date) {
      newRows.push(row);
    }
  }
  return newRows;
}

/** Main logic: fetch missing rates from ECB and insert into Postgres. */
export async function run(): Promise<FetcherResult> {
  const ecbCurrencies = determineEcbCurrencies(ECB_CURRENCIES);
  const dateRanges = await getRateDateRanges(ECB_CURRENCIES);

  // Only a currency without any row needs the full history; otherwise fetch forward
  // from what is already stored, so a daily run covers one or two days.
  const needsBackfill = ECB_CURRENCIES.some((c) => !(c in dateRanges));

  let start: string;
  if (needsBackfill) {
    start = ECB_EARLIEST_DATE;
  } else {
    const earliestMax = Object.values(dateRanges)
      .map((r) => r.max_date)
      .sort()[0];
    start = addDays(earliestMax, 1);
  }

  const end = todayIso();

  if (start > end) {
    console.log("All ECB rates are up to date");
    return { inserted: 0, latest_date: end, missing_currencies: [] };
  }

  const csvText = await fetchEcbRates(ecbCurrencies, start, end);
  const ecbRates = parseEcbCsv(csvText);

  if (ecbRates.length === 0) {
    // ECB published nothing in this period: a weekend, a TARGET holiday, or a run
    // before the ~16:00 CET publication. Nothing is missing and nothing is suspended,
    // so the stored coverage is reported unchanged and no warning is raised.
    console.log(`No ECB observations for period ${start} to ${end}`);
    const storedMaxDates = Object.values(dateRanges).map((r) => r.max_date).sort();
    const latestStored =
      storedMaxDates.length > 0 ? storedMaxDates[storedMaxDates.length - 1] : end;
    return { inserted: 0, latest_date: latestStored, missing_currencies: [] };
  }

  const allRows = convertEurRatesToUsd(ecbRates);

  const returnedCurrencies = new Set(allRows.map((r) => r.base_currency));
  const missing = ECB_CURRENCIES.filter((c) => !returnedCurrencies.has(c));
  if (missing.length > 0) {
    console.warn(
      `ECB did not return rates for currencies: ${missing.join(", ")} (they may be suspended)`,
    );
  }

  const newRows = filterNewRows(allRows, dateRanges, ECB_CURRENCIES);
  const inserted = await insertRows(newRows);

  const latestInserted =
    newRows.length > 0
      ? newRows.map((r) => r.rate_date).sort().reverse()[0]
      : end;

  console.log(`ECB: inserted ${inserted} rows, latest date: ${latestInserted}`);

  return { inserted, latest_date: latestInserted, missing_currencies: missing };
}
